### PR TITLE
wg_engine: multicanvas support

### PR DESCRIPTION
**Pull request by SergeyLebedkin**
Added ability for canvas to render into external target texture, in addition to surface 
Issue https://github.com/thorvg/thorvg/issues/2745

See MultyCanvas example
image

### CODE REVIEW

File: inc/thorvg.h [hunk 1]
```diff
@@ -1743,6 +1743,25 @@ class TVG_API WgCanvas final : public Canvas
      */
     Result target(void* instance, void* surface, uint32_t w, uint32_t h, void* device = nullptr) noexcept;
```
Comment (hermet) [inline inc/thorvg.h:1744]: @SergeyLebedkin I think we can refine API with this one.

```
 Result target(device, instance, surface / texture, w, h, int type (0: surface / 1: texture));
```

Comment (hermet) [reply inc/thorvg.h:1744]: @SergeyLebedkin please separate engine / example patches.

File: inc/thorvg.h [hunk 2]
```diff
@@ -1741,7 +1742,7 @@ class TVG_API WgCanvas final : public Canvas
      * @see Canvas::viewport()
      * @see Canvas::sync()
      */
-    Result target(void* instance, void* surface, uint32_t w, uint32_t h, void* device = nullptr) noexcept;
+    Result target(void* instance, void* surface, uint32_t w, uint32_t h, int type = 0, void* device = nullptr) noexcept;
```
Comment (hermet) [inline inc/thorvg.h:1745]: I'd like to suggest parameter order and names: (device, instance, target, w, h, type = 0)

File: inc/thorvg.h [hunk 3]
```diff
@@ -1728,9 +1728,10 @@ class TVG_API WgCanvas final : public Canvas
      * @brief Sets the drawing target for the rasterization.
      *
      * @param[in] instance WGPUInstance, context for all other wgpu objects.
-     * @param[in] surface WGPUSurface, handle to a presentable surface.
+     * @param[in] surface/texture WGPUSurface/WGPUTexture, handle to a presentable surface/texture.
```
Comment (hermet) [inline inc/thorvg.h:1731]: `Either WGPUSurface or WGPUTexture, serving as handles to a presentable surface or texture.`

File: src/renderer/tvgWgCanvas.cpp [hunk 4]
```diff
@@ -52,25 +52,33 @@ WgCanvas::~WgCanvas()
 {
 #ifdef THORVG_WG_RASTER_SUPPORT
     auto renderer = static_cast<WgRenderer*>(Canvas::pImpl->renderer);
-    renderer->target(nullptr, 0, 0);
+    renderer->target(nullptr, (WGPUSurface)nullptr, 0, 0, nullptr);
```
Comment (hermet) [inline src/renderer/tvgWgCanvas.cpp:55]: ` (WGPUSurface)nullptr` ->  `nullptr` ?

File: src/renderer/tvgWgCanvas.cpp [hunk 5]
```diff
@@ -52,25 +52,33 @@ WgCanvas::~WgCanvas()
...
+    if (!instance || !target || (w == 0) || (h == 0)) return Result::InvalidArguments;
 
     //We know renderer type, avoid dynamic_cast for performance.
     auto renderer = static_cast<WgRenderer*>(Canvas::pImpl->renderer);
     if (!renderer) return Result::MemoryCorruption;
 
-    if (!renderer->target((WGPUInstance)instance, (WGPUSurface)surface, w, h, (WGPUDevice)device)) return Result::Unknown;
+    // surface
+    if (type == 0) {
+        if (!renderer->target((WGPUInstance)instance, (WGPUSurface)target, w, h, (WGPUDevice)device)) return Result::Unknown;
```
Comment (hermet) [inline src/renderer/tvgWgCanvas.cpp:75]: You can cast the target in the renderer engine, not in the canvas. There's no need for an if-else here.

**Review by hermet — CHANGES_REQUESTED** [review]
0. Your patches are messed up now. please submit PR with 2 commits:
 a. `engine change + broken examples by api changes` 
 b. `multi-canvas example` 
1. please check minor comments
2. please update the tvg_wgcanvas_set_target() in `capi`
3. please update api call from the `tvgWasmLottieAnimation.cpp`

**Comment by hermet** [comment]
@SergeyLebedkin Please double-check this memory sanitizer report; it occurs in every example.
```
![image](https://github.com/user-attachments/assets/52e482b9-1dff-4c20-abe2-566f53ef8c3e)
```

**Review by tinyjin — APPROVED** [review]
CleanShot 2024-11-20 at 13 58 34@2x

CleanShot 2024-11-20 at 14 04 48@2x

File: examples/MultiCanvas.cpp [hunk 6]
```diff
@@ -303,28 +325,77 @@ void runWg()
...
+    wgpuTextureDestroy(renderTarget);
+    wgpuTextureRelease(renderTarget);
+    wgpuDeviceRelease(device);
+    wgpuAdapterRelease(adapter);
+    wgpuSurfaceUnconfigure(surface);
+    wgpuSurfaceRelease(surface);
     wgpuInstanceRelease(instance);
     SDL_DestroyWindow(window);
 #else
     cout << "webgpu driver is not detected!" << endl;
```
Comment (hermet) [inline examples/MultiCanvas.cpp:397]: Please retain the original message. This doesn't mean the driver problem, user might build without thorvg webgpu backend. 
```
cout << "Not Support WebGPU" << endl;
```

Comment (SergeyLebedkin) [reply examples/MultiCanvas.cpp:397]: fixed

File: examples/MultiCanvas.cpp [hunk 7]
```diff
@@ -303,27 +325,76 @@ void runWg()
...
-        tvgexam::verify(canvas->target(instance, surface, WIDTH, HEIGHT));
+        tvgexam::verify(canvas->target(device, instance, renderTarget, SIZE, SIZE, 1));
 
         content(canvas.get());
         if (tvgexam::verify(canvas->draw())) {
             tvgexam::verify(canvas->sync());
         }
-    }
 
+        // After the GlCanvas::sync() function, the content is rendered to the framebuffer
```
Comment (hermet) [inline examples/MultiCanvas.cpp:375]: an incorrect message.

**Comment by hermet** [comment]
@SergeyLebedkin Please double check SVG example
image

**Comment by SergeyLebedkin** [comment]
> @SergeyLebedkin Please double check SVG example
@hermet can`t repoduce it on SVG and Lottie examples

Build script:
`meson setup builddir -Dloaders=all -Dexamples=true -Db_sanitize="address,undefined" -Dengines="sw, wg, gl`

**Comment by hermet** [comment]
> > @SergeyLebedkin Please double check SVG example
> 
> @hermet can`t repoduce it on SVG and Lottie examples
> 
> Build script: `meson setup builddir -Dloaders=all -Dexamples=true -Db_sanitize="address,undefined" -Dengines="sw, wg, gl`

on my side, it still occur the issue, curlprit -> `matheusmdx.svg` 
image

**Comment by hermet** [comment]
@SergeyLebedkin Another memory violation report came out after the update:

Please check `GradientMasking`
```
Indirect leak of 160 byte(s) in 1 object(s) allocated from:
    #0 0x7fd9a3c99887 in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x7fd9a0681b62 in wgpu_core::resource::Buffer$LT$A$GT$::destroy::h686c20d29f6f7998 (/usr/local/lib/libwgpu_native.so+0x24eb62)
    #2 0x7fd9a1c622fa in WgMeshData::release(WgContext&) ../src/renderer/wg_engine/tvgWgRenderData.cpp:128
    #3 0x7fd9a1c3211f in WgCompositor::release(WgContext&) ../src/renderer/wg_engine/tvgWgCompositor.cpp:58
    #4 0x7fd9a1cd4cd2 in WgRenderer::release() ../src/renderer/wg_engine/tvgWgRenderer.cpp:58
    #5 0x7fd9a1cd5b1d in WgRenderer::~WgRenderer() ../src/renderer/wg_engine/tvgWgRenderer.cpp:32
    #6 0x7fd9a1cd5b1d in WgRenderer::~WgRenderer() ../src/renderer/wg_engine/tvgWgRenderer.cpp:33
    #7 0x7fd9a18e8a76 in tvg::Canvas::Impl::~Impl() ../src/renderer/tvgCanvas.h:50
    #8 0x7fd9a18e8a76 in tvg::Canvas::~Canvas() ../src/renderer/tvgCanvas.cpp:36
    #9 0x7fd9a197b4d5 in tvg::WgCanvas::~WgCanvas() ../src/renderer/tvgWgCanvas.cpp:58
    #10 0x563a3a2a12fe in tvgexam::Window::~Window() ../examples/Example.h:160
    #11 0x563a3a2a12fe in tvgexam::WgWindow::~WgWindow() ../examples/Example.h:430
    #12 0x563a3a2a12fe in tvgexam::WgWindow::~WgWindow() ../examples/Example.h:430
    #13 0x563a3a290583 in std::default_delete<tvgexam::Window>::operator()(tvgexam::Window*) const /usr/include/c++/11/bits/unique_ptr.h:85
    #14 0x563a3a290583 in std::unique_ptr<tvgexam::Window, std::default_delete<tvgexam::Window> >::~unique_ptr() /usr/include/c++/11/bits/unique_ptr.h:361
    #15 0x563a3a290583 in tvgexam::main(tvgexam::Example*, int, char**, unsigned int, unsigned int, unsigned int, bool) ../examples/Example.h:527
    #16 0x563a3a296227 in main ../examples/GradientMasking.cpp:154
    #17 0x7fd99f9b0d8f in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
```

**Comment by mgrudzinska** [comment]
macOs (intel) - no info from sanitizers: address, undef

**Comment by SergeyLebedkin** [comment]
> @SergeyLebedkin Another memory violation report came out after the update, please double-check them.
> `Accessor`
> `AnimateMasking`

> macOs (intel) - no info from sanitizers: address, undef

Also can`t reproduce it on wsl

File: inc/thorvg.h [hunk 8]
```diff
@@ -1788,7 +1789,7 @@ class TVG_API WgCanvas final : public Canvas
      * @see Canvas::viewport()
      * @see Canvas::sync()
      */
-    Result target(void* instance, void* surface, uint32_t w, uint32_t h, void* device = nullptr) noexcept;
+    Result target(void* device, void* instance, void* surface, uint32_t w, uint32_t h, int type = 0) noexcept;
```
Comment (hermet) [inline inc/thorvg.h:1792]: `surface` -> `target`

File: inc/thorvg.h [hunk 9]
```diff
@@ -1774,11 +1774,12 @@ class TVG_API WgCanvas final : public Canvas
     /**
      * @brief Sets the drawing target for the rasterization.
      *
+     * @param[in] device WGPUDevice, a desired handle for the wgpu device. If it is @c nullptr, ThorVG will assign an appropriate device internally.
      * @param[in] instance WGPUInstance, context for all other wgpu objects.
-     * @param[in] surface WGPUSurface, handle to a presentable surface.
+     * @param[in] surface or texture either WGPUSurface or WGPUTexture, serving as handles to a presentable surface or texture.
```
Comment (hermet) [inline inc/thorvg.h:1779]: `target Either WGPUSurface or WGPUTexture, serving as handles to a presentable surface or texture.`

File: src/bindings/capi/thorvg_capi.h [hunk 10]
```diff
@@ -560,19 +560,20 @@ TVG_API Tvg_Canvas* tvg_wgcanvas_create(void);
...
+* @param[in] type 0: surface, 1: texture are used as pesentable target
 *
 * \return Tvg_Result enumeration.
 * \retval TVG_RESULT_INSUFFICIENT_CONDITION if the canvas is performing rendering. Please ensure the canvas is synced.
 * \retval TVG_RESULT_NOT_SUPPORTED In case the wg engine is not supported.
 *
 * \note Experimental API
 */
-TVG_API Tvg_Result tvg_wgcanvas_set_target(Tvg_Canvas* canvas, void* instance, void* surface, uint32_t w, uint32_t h, void* device);
+TVG_API Tvg_Result tvg_wgcanvas_set_target(Tvg_Canvas* canvas, void* device, void* instance, void* surface, uint32_t w, uint32_t h, int type = 0);
```
Comment (hermet) [inline src/bindings/capi/thorvg_capi.h:576]: surface -> target

File: src/bindings/capi/thorvg_capi.h [hunk 11]
```diff
@@ -560,19 +560,20 @@ TVG_API Tvg_Canvas* tvg_wgcanvas_create(void);
...
 * \brief Sets the drawing target for the rasterization.
 *
-* \param[in] instance WGPUInstance, context for all other wgpu objects.
-* \param[in] surface WGPUSurface, handle to a presentable surface.
-* \param[in] w The width of the surface.
-* \param[in] h The height of the surface.
-* \param[in] device WGPUDevice, a desired handle for the wgpu device. If it is @c nullptr, ThorVG will assign an appropriate device internally.
+* @param[in] device WGPUDevice, a desired handle for the wgpu device. If it is @c nullptr, ThorVG will assign an appropriate device internally.
+* @param[in] instance WGPUInstance, context for all other wgpu objects.
+* @param[in] surface or texture either WGPUSurface or WGPUTexture, serving as handles to a presentable surface or texture.
```
Comment (hermet) [inline src/bindings/capi/thorvg_capi.h:565]: target Either WGPUSurface or WGPUTexture, serving as handles to a presentable surface or texture.

**Review by hermet — COMMENTED** [review]
@SergeyLebedkin please polish up the api doc / parameter

**Review by hermet — APPROVED** [review]

**Pull request merged**